File: android/src/main/java/expo/modules/rustysearch/ExpoRustySearchModule.kt
package expo.modules.rustysearch

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoRustySearchModule : Module() {

  // JNI functions return String directly
  private external fun initializeIndex(): String
  private external fun clearIndex(): String
  private external fun addDocument(title: String, body: String): String
  private external fun addDocumentsBulk(json: String): String
  private external fun searchDocuments(query: String): String
  private external fun getDocumentCount(): String

  companion object {
    init {
      System.loadLibrary("expo_search_core")
    }
  }

  override fun definition() = ModuleDefinition {
    Name("ExpoRustySearch")

    Events("onChange")

    Function("hello") {
      "Hello world!"
    }

    AsyncFunction("initializeIndex") {
      initializeIndex()
    }

    AsyncFunction("clearIndex") {
      clearIndex()
    }

    AsyncFunction("addDocument") { title: String, body: String ->
      addDocument(title, body)
    }

    AsyncFunction("addDocumentsBulk") { json: String ->
      addDocumentsBulk(json)
    }

    AsyncFunction("searchDocuments") { query: String ->
      searchDocuments(query)
    }

    AsyncFunction("getDocumentCount") {
      getDocumentCount()
    }
  }
}

File: ios/ExpoRustySearchModule.swift
import ExpoModulesCore

public class ExpoRustySearchModule: Module {
  // Helper to handle Rust string pointers safely
  func rustToString(_ ptr: UnsafePointer<Int8>?) -> String {
    guard let ptr = ptr else { return "" }
    let result = String(cString: ptr)
    free_string(UnsafeMutablePointer(mutating: ptr))
    return result
  }

  public func definition() -> ModuleDefinition {
    Name("ExpoRustySearch")

    Constant("PI") {
      Double.pi
    }

    Events("onChange")

    Function("hello") {
      return "Hello world! 👋"
    }

    AsyncFunction("setValueAsync") { (value: String) in
      self.sendEvent("onChange", [
        "value": value
      ])
    }

    AsyncFunction("initializeIndex") { () -> String in
      return self.rustToString(initialize_index())
    }

    AsyncFunction("clearIndex") { () -> String in
      return self.rustToString(clear_index())
    }

    AsyncFunction("addDocument") { (title: String, body: String) -> String in
      return self.rustToString(add_document(title, body))
    }

    AsyncFunction("addDocumentsBulk") { (json: String) -> String in
      return self.rustToString(add_documents_bulk(json))
    }

    AsyncFunction("searchDocuments") { (query: String) -> String in
      return self.rustToString(search_documents(query))
    }

    AsyncFunction("getDocumentCount") { () -> String in
      return self.rustToString(get_document_count())
    }
  }
}